feat(clients): add transaction support

ibilux opened the pull request:
Hello,

This aims to add transaction support to Trailbase clients:

- Tried keeping the changes to minimal and close as possible with each client’s existing coding style.
- Added transaction tests to each client library and all tests seem to pass.
- Moved Go client tests above the Swift client tests in `.pre-commit-config.yaml` because Swift tests were taking forever to complete.
- I did **not** bump any client versions yet.

The code definitely should be reviewed (because even thought all the high-level programming languages have some kind of an abstract similarities, but each one may have some language-specific improvements or performance idioms).

I can see that you are currently focused on `wasm` so this can wait whenever you have time for it.

Cheers,

ibilux committed:
wip: add transaction support to clients

Signed-off-by: Bilux <i.bilux@gmail.com>

ibilux committed:
wip: enable record transactions for tests

Signed-off-by: Bilux <i.bilux@gmail.com>

ibilux committed:
wip: clean up some clients transaction support code

Signed-off-by: Bilux <i.bilux@gmail.com>

ignatz commented:
> I was hoping you might provide some compelling arguments as you usually do 😅

Appreciated

>  you always think ahead of all possible outcomes.

I wish :sweat_smile: 

> I think we shouldn't try to squeeze both records and transactions into the same surface. It feels cleaner to separate them into their own dedicated namespaces.

I think that's a good goal. To tally up all discussed options that meet this goal (I guess the previous (2) and (3) wouldn't fit) and avoid Transaction-object lifecycle concerns:

1. Have a symmetric `transactions("name")`/`ops("name")` API mirroring `records("name")`
2. Have a `client.commit([client.transaction() .records("users").create(record)])` API (your latest proposal).
3. Have operations always be deferred and add an ~`execute` method.

I'm going to say that (1) and (2) are virtually the same if you fold `.transaction().records("name")` into `.transaction("name")`. Maybe you had some use-case in mind?

I've been careful to push to hard for (3) because it would be the biggest change (not even that much more work but also for users).  I do think, it's probably the most intuitive, i.e. you build and operation and then you either execute it right away or you add it to a transaction. Rather than using two different APIs.

Similarly, there's currently a bulk creation, which isn't a transaction. We could do away with it entirely and have an `await client.execute([op, ... ], { transaction = false})` API. In other words we'd have fewer APIs and just different execution modes: single, bulk, transaction.

ignatz commented:
> Ultimately, I think we need users opinion and feedback on this before finalizing the approach ?

That would be ideal but is sadly hard to come by. I'm very open to pull folks in, any suggestions?

> I’m not considering myself a user for now to avoid bias 😅

I think you should. You're taking all this time out of your day to look at this problem and combined with your experience, your opinion matters (a lot) and is probably more formed than some just responding to a survey on the toilet :)

(Just as a reminder, we aren't blocked. We can decide. More eyeball more better when available)

> I think immediate execution should be named execute or run, because query typically implies read-only operations.

I see where you're coming from and am happy either way. (In the context of databases, `client.query` usually returns records and can be mutating, while `client.execute` is typically a fire-and-forget and therefore tends to be mutating).

> And Since we’re already going through major changes, introducing a create just to remove it later feels like another breaking change. If we postpone non-transactional bulk operations, I think bulk is a better name and should be typed to accept only create operations.

Feedback taken. Then let's update the transaction handler to support both execution modes. (Happy to give it a go, just let me know)

ignatz commented:
Hey @ibilux, sorry for the slow reply. Wasn't feeling so hot but now back on my feet.

## Client

I love your proposed client-side API - very clean :heart: . As for your points on asymmetry, I could again be swayed either way. Personally, I don't mind that read/list are deferred too much. To riff a bit more on your proposal, if you moved expand into the `.query()`, you could even have:

```ts
const read : ReadOperation<Simple> = api.read(id);
const simple = await read.query({ expand: ["author"] });

const stream = read.subscribe();
```

On the bright side, it would remove the asymmetry but it may  make subscribe a bit less discoverable. Also none of the other `query` methods have any arguments. Not sure, which is best.

Spelling it out here, I feel that `subscribe` is more like `list` than `read` however it doesn't support fancy filters (it wouldn't be easy to implement but it would make a lot of sense for it to do). It also doesn't support `expand` (which maybe it should, but I'm feeling less strongly about). Arguing against myself, I'm not sure folding it into either list or read deferred operation would be great, at least at this point.

To sum up:

1. read/list deferred, no arguments in query and a separate subscribe.
2. read/list immediate
3. arguments on `read/list(id).query` with separate subscribe

None of them strikes me as "terrible". From a mere symmetry argument, I'm actually leaning (1). Thinking further, maybe `subscribe` should be deferred as well....
`subscribe` is fairly minimal right now. I could totally see a world in which subscribe would or should support a more complex lifecycle, e.g. pause, resume, onDisconnect. Especially the latter, it would be nice to provide a mechanism to "catch back up" when the connection is interrupted or the backend is restarted (e.g. for an update).

## Server

> For the bulk operation on the server side, is the idea simply to iterate through the operations like this?
Or do you think we should aim for something more like run_queries

Still paging in, where do you see the main difference between your sketch and `run_queries`?

ibilux committed:
Revert

ibilux committed:
wip: add transaction support to `js` client

Signed-off-by: Bilux <i.bilux@gmail.com>

ibilux commented:
> Do you feel like we're happy enough and at a point where we could jump to implementation? FWIW, I'm very exited how this is shaping up 🙏

I have updated the code for `js` client library only to see how it go. I noticed from the github action that the examples should be updated too after we settle on the implementation.
The bulk creation should be removed from `create_record.rs` too I guess ?

> ## Server
> Still paging in, where do you see the main difference between your sketch and `run_queries`?

From may understanding to the code (correct me if I'm wrong) the `run_queries` function has a file handling that ensures files are written to object storage before the database transaction commits, and if the transaction fails, the file manager automatically cleans up any files that were written ?

Shouldn't a similar approach be implemented for transaction API ?

ignatz commented:
> I have updated the code for js client library only to see how it go.

You're a machine :clap:

> should be updated too after we settle on the implementation.

After writing it out, how do you feel about it?

I would love to hear your unbiased feedback. Yet, I'll share my thoughts to minimize round-trips... I love the implementation. Everything is extremely symmetric, composable and mockable. Yet, seeing it spelled out it's clearly more cumbersome then before when you're not using transactions. I'm wondering if we could just leave the implementation as is and add short-hands, e.g. `create: (props) => await this.createOp(props).query()`. We'd have two sets of APIs, perfectly symmetric with a single implementation (it would also end up being backwards compatible).  WDYT? (I'm sorry for going back and forth on this :pray: - I really do love the current implementation)

> From may understanding to the code (correct me if I'm wrong) the run_queries function has a file handling that ensures files are written to object storage before the database transaction commits, and if the transaction fails, the file manager automatically cleans up any files that were written ?
>
> Shouldn't a similar approach be implemented for transaction API ?

Yikes, if that's not already the case, I would consider it a bug. You're absolutely right that also for transaction we need to cleanup overriden files or files after a transaction fails. I'll take a look.

ibilux commented on crates/assets/js/client/src/index.ts:
```diff
@@ -186,54 +186,146 @@ export type Or = {
-      expand?: string[];
-    },
-  ): Promise<T>;
+export interface Operation {
```
**We can use Types:**
```ts
interface CreateOperation {
  Create: { api_name: string; value: Record<string, unknown> };
}
interface UpdateOperation {
  Update: { api_name: string; record_id: string | number; value: Record<string, unknown> };
}
interface DeleteOperation {
  Delete: { api_name: string; record_id: string | number };
}

type Operation = CreateOperation | UpdateOperation | DeleteOperation;

export class CreateOperation<T>  implements DeferredMutation<string | number>
{
  toJSON(): CreateOperation { ... }
}
```

---

**Or use generic helper:**

```ts
type OperationKind = "Create" | "Update" | "Delete";

type OperationMap = {
  Create: { api_name: string; value: Record<string, unknown> };
  Update: { api_name: string; record_id: string | number; value: Record<string, unknown> };
  Delete: { api_name: string; record_id: string | number };
};

type Operation<K extends OperationKind> = { [P in K]: OperationMap[P] };

export class CreateOperation<T>  implements DeferredMutation<string | number>
{
  toJSON(): Operation<"Create"> { ... }
}
```

ibilux replied:
You’re right, the current `Operation` interface allows `AnyOfOrNone`, which means you could accidentally construct an invalid shape (none or multiple at once). Switching to a union (`CreateOp `| `UpdateOp` | `DeleteOp`) enforces exactly one and avoids that issue.

ibilux committed:
wip: update `js` client api

ignatz commented on crates/assets/js/client/src/index.ts:
```diff
@@ -186,54 +186,154 @@ export type Or = {
+    private readonly apiName: string,
+    private readonly record: Partial<T>,
+  ) {}
+  async query(): Promise<string> {
```
Should this match the T in `DeferredMutation<T>`?

ignatz commented on crates/assets/js/client/src/index.ts:
```diff
@@ -275,71 +375,110 @@ export class RecordApiImpl<T = Record<string, unknown>>
+    expand?: string[];
+  }): ListOperation<T>;
+
+  readOp(
```
:super nitty: just from a discoverability point of view, i'd group <operation> and <operation>Op next to each other.

ignatz commented on crates/assets/js/client/src/index.ts:
```diff
@@ -275,71 +375,110 @@ export class RecordApiImpl<T = Record<string, unknown>>
+    private readonly client: Client,
+    private readonly name: string,
+  ) {
+    this._path = `${recordApiBasePath}/${this.name}`;
```
I don't think this is needed anymore :tada:

ibilux replied:
It was used in `subscribe` method, but it’s better to remove and replace it.

ignatz reviewed (changes requested):
This looks phenomenal - only minor nits (the accidental looking changes to .pre-commit-config.yaml and config.textproto are probably just stale?).

I would love to merge this. Thank you so much, both for the work and bearing with me :pray:

ibilux commented on crates/assets/js/client/src/index.ts:
```diff
@@ -186,54 +186,154 @@ export type Or = {
-    count?: boolean;
-    expand?: string[];
-  }): Promise<ListResponse<T>>;
+export interface CreateOp {
```
I was planning to ask you after we settled on the implementation about which classes should remain exported and which should not.
I’ve removed the export. But, I think we should keep `DeferredOperation` and `DeferredMutation` only as a reference for other clients implementation.

ibilux commented on crates/assets/js/client/src/index.ts:
```diff
@@ -186,54 +186,154 @@ export type Or = {
+    record_id: string | number;
+  };
+}
 
```
Agreed, inlining makes it cleaner.

ibilux committed:
wip: update js client api

ibilux commented:
> This looks phenomenal - only minor nits (the accidental looking changes to .pre-commit-config.yaml and config.textproto are probably just stale?).
> 
> I would love to merge this. Thank you so much, both for the work and bearing with me 🙏

Thanks @ignatz! I’m really glad I could help, and I’ve enjoyed our conversations along the way.
I’ve addressed all your points.

Regarding the next steps, would it make sense to close this PR and open separate ones to keep the git history cleaner? I’m thinking:

1. Adding the bulk operation to `transaction.rs`
2. Enabling the transactions API in `client/testfixture/config.textproto` (unless you have a preferred approach)
3. Updating the JS client

ignatz commented:
> I’m really glad I could help

Much much much appreciated :pray:

> I’ve enjoyed our conversations along the way

pfew :sweat_smile: 

> Regarding the next steps, would it make sense to close this PR and open separate ones to keep the git history cleaner? I’m thinking:
>
>    Adding the bulk operation to transaction.rs
>    Enabling the transactions API in client/testfixture/config.textproto (unless you have a preferred approach)
>    Updating the JS client

I leave this 100% up to you. Ultimately it's just cosmetics and doesn't take away from the quality of your work. Just let me know.

(I won't be available this weekend, only bringing this up in case you're keen to get this in before. Want to be mindful of your time)

ibilux committed:
Finale clean up

ignatz commented:
Quick update: I published a new npm package for the client and cutting a new server release :

ignatz commented:
> Anyway, I'm sure this API will be as good as the rest of TrailBase. I'll use it with TanStack and SolidJS to develop an multiuser SPA with client-first collections (multi-database support would be great for this).

Curious question, since I'm currently looking into multi-DB. Are you thinking of a small, more-or-less fixed number of DBs or a 1:1 DB to user mapping (~multi-tenancy)? The former is pretty close, the latter will need a bit more thought.

ignatz commented:
> Yes, a 1:1 DB per customer. Each customer can have zero or more users accessing a synchronized copy of the same BD. Migrations will undoubtedly be more complex, but I think there will be improvements in simplicity and overall performance. Something similar to what [Electric's Shapes](https://electric-sql.com/product/electric#how-does-it-work) offer, but with SQLite DBs.

I'm no expert on Electric's offering, maybe you can help me out, it looks like there's a single central DB with some policy engine that decides which changes get streamed out to which on-device databases.

What I first understood and am working on is: multiple SQLite DBs on the server. From multi-tenancy I would expect data silos, i.e. users can only ever access (read & write) data in their silos (e.g. 1:1 mapping for users or something more complex like members of a customer, as you're asking).

FWIW, something like Electric's policy-based streaming one can already model on the existing building blocks. Specifically, you can have a group membership table to associate records in another table with a group. Then you can setup access rules that only group members can access said records, in which case any user subscribing for changes will only see records they have access to.

My intuition is that one always wants to have more fine-grained access on streaming then just, stream this DB. Even in a multi-tenancy world, you might only want certain updates.

Note that TB subscriptions also support client-side filters to further narrow the changes, this is on-top of the server-side access control. For example, you could subscribe to a stream of geo coordinates associated with your customer, which are within a specific bounding box.

Maybe you could talk a bit more to what you're trying to do to then work our way backwards and how to best model it.

aMarCruz commented:
> I'm no expert on Electric's offering, maybe you can help me out, it looks like there's a single central DB with some policy engine that decides which changes get streamed out to which on-device databases.
>

Yes, it is a single DB.
[Shapes are defined by](Shapes are defined by:

a [table](https://electric-sql.com/docs/guides/shapes#table), such as items
an optional [where clause](https://electric-sql.com/docs/guides/shapes#where-clause) to filter which rows are included in the shape
an optional [columns](https://electric-sql.com/docs/guides/shapes#columns) clause to select which columns are included
A shape contains all of the rows in the table that match the where clause, if provided. If a columns clause is provided, the synced rows will only contain those selected columns.

I'm not an expert in Electric, but the concept is simple: they're SQL queries that expose a set of records to the client as if it were the entire table. It's what some RDBMSs called "Views", but mutables.

> What I first understood and am working on is: multiple SQLite DBs on the server. From multi-tenancy I would expect data silos, i.e. users can only ever access (read & write) data in their silos (e.g. 1:1 mapping for users or something more complex like members of a customer, as you're asking).
> 

Yes, 1 SQLite DB for customer (the root user for this db).
This is similar to Electric's Shapes, but without putting all your eggs in one only basket, with the benefits that this entails :-)

> FWIW, something like Electric's policy-based streaming one can already model on the existing building blocks. Specifically, you can have a group membership table to associate records in another table with a group. Then you can setup access rules that only group members can access said records, in which case any user subscribing for changes will only see records they have access to. At least on Web you can use TanStack DB as a client-side queryable, synced DB.
> 
> My intuition is that one always wants to have more fine-grained access on streaming then just, stream this DB. Even in a multi-tenancy world, you might only want certain updates.
> 
> Note that TB subscriptions also support client-side filters to further narrow the changes, this is on-top of the server-side access control. For example, you could subscribe to a stream of geo coordinates associated with your customer, which are within a specific bounding box.
> 
> Maybe you could talk a bit more to what you're trying to do to then work our way backwards and how to best model it.

My app will be like any other administrative App, think of one for restaurants, for example...

A client registers their restaurant in the App. At that moment, a new DB is created on the server. This DB will act as a backup and synchronization point for all restaurant users (cashiers, POS systems, accountants, etc.). The app needs to know the DB URL and whether the subscription is active. The other restaurants are not accessible to the App. This simplifies everything and minimizes database locks.

On the server, a master DB (not directly accessible from the app) will contain records of restaurant owners, their account type, subscribed modules, balance, etc. The database filename corresponds to the owner's ID and maintains a read-only record with their data in a table.

**On the server**, it's only necessary to create the new DB when a restaurant registers and grant access to its owner. The rest of the users and their roles in that restaurant are controlled by the App.

Of course, this idea isn't new. Over 20 years ago, I created a similar app in Visual FoxPro 6 with local DBFs synchronized with an NT4 server, and it worked very well. Implementing it with MS SQL would have taken us twice as much time and money.

Sorry for rambling on. I hope I answered correctly; English isn't my first language, and I barely understand it.

aMarCruz commented:
> > My understanding was that ‘multi-database’ in TrailBase meant multiple SQLite DBs, each with its own schema. And I was curious how the separation of these DBs would work in terms of selection, migration, storage, ... etc.
> 
> Currently, I'm aiming for the minimal viable product to get things going and iron out the kinks. Specifically, allowing to register additional databases, which map to `<traildepot>/data/<name>.db` and `<traildepot>/migrations/<name>/*.sql`, which means you can have independent schemas. Record APIs can then reference fully-qualified tables, e.g. `<name>.mytable`.
> 
> I'm currently also attaching all DBs to the same connection, which has the benefit that cross-DB `VIEW`s will simply work, however at the downside of only supporting up to 124 DBs. To allow more DBs, we'd have to manage connections more actively, e.g. manage different connections or dynamically attach/detach DBs.
>

This is one of the problems to be solved by the implementation; I can imagine a successful app requiring 1,000 DBs... or more.
 
> EDIT: spelling things out, I'm starting to think that cross-DB VIEWs is maybe a feature that can wait in favor of more independent DBs 🤷‍♀️
> 
> > I think multi-tenancy could already be handled on top of this using TrailBase **`ACL`** and **`access rules`** ?
>
> Above limitations aside, it depends what you expect it to do. If you have a 1000 tenants, every API would become 1000 APIs if they needed to be backed by independent DBs. If you want the same schema for each tenant, you'd also need some bootstrapping and maybe a few more lifecycle hooks, e.g. to clean up deleted tenants.
> 

One App, one schema.

aMarCruz commented:
> Excuse my intrusion, but I’ve been looking forward to multi-database support since I saw the roadmap. My understanding was that ‘multi-database’ in TrailBase meant multiple SQLite DBs, each with its own schema. And I was curious how the separation of these DBs would work in terms of selection, migration, storage, ... etc. I think multi-tenancy could already be handled on top of this using TrailBase **`ACL`** and **`access rules`** ?

Multiple databases with the same schema and migrations.
But it's a specific requirement; I think it's suitable for SQLite, although perhaps not common.

ignatz commented:
> Sorry for rambling on. I hope I answered correctly; English isn't my first language, and I barely understand it.

Same boat. Your english is great. Your insights and 20 year war stories are greatly appreciated.

Everything you say makes a lot of sense. I'm mostly just trying to keep "multi-tenancy" and "syncing" as **two separate features**, with possibly synergistic effects. Specifically, Electric with shapes seems to market itself mostly on read-only syncing from a single master Postgres instance.  (I did find some reference to support for multiple but independent PG DBs in code: https://github.com/electric-sql/electric/commit/3775f1c23b11b96eafa35e0ce14c698038e8d8bf - but not when skimming the docs).

## Syncing

> On the server, it's only necessary to create the new DB when a restaurant registers and grant access to its owner. The rest of the users and their roles in that restaurant are controlled by the App.

Skimming Electric's docs, this is the one part where I'm not sure. Please correct me if I'm wrong. My understanding is that the docs use "client" and "app" pretty interchangeably. So, a sync-client might exist either in one of your backends or may be directly running on your users' devices as part of an app. Either way, the new DB is created on the client side and not within the Electric's stack (PG + Sync). This client-side DB may then be backed by SQLite, PGLite, TanStack DB, ... . In other words, whether that per-client (which may correspond to a "tenant") DB is "on the server" (as you say) or somewhere else, depends on your setup and is not Electric's responsibility.... then again, I may be in the weeds.

If this understanding is roughly correct, many similar building blocks are already in place. Specifically, you can sync your table contents into TanStack DB, whether it's running on the server or in your web app.

(Btw. I'm sure Electric is great, I'm not trying to trash talk, I'm mostly trying to understand and then think about what aspects or features similar to or different from prior art may make sense in TrailBase).

I'm wondering if maybe you're thinking ahead, e.g. unlike Electric, breaking up per-tenant data into separate DBs could allow for a more 1:1 sync between a server-side DB shard and a client? Maybe even enable write-path syncing?

## Multi-Tenancy

> Multiple databases with the same schema and migrations.

This is what I would understand as multi-tenancy, too. IIUC, Electric doesn't help with this either?

Relatedly, you could also have a multi-environment setup, e.g. test, QA, and prod data physically separated, as opposed to data for different tenants physically separated.

Either way, either would be cool to support. As you point out there's some extra work for lifecycle management.

The first step right now for me is to support multiple independent databases to allow different data to be kept in different physical locations and overcome some limitations (e.g. locking). This wouldn't yet allow to auto-magically route different tenants to different DBs but you could manually break by topic or establish a manual routing based on policy requirements (e.g. EMEA vs north-america :woman_shrugging:).

> I can imagine a successful app requiring 1,000 DBs or more, although not all connected at the same time.

Absolutely and will be an absolute requirement for true multi-tenancy... baby steps :)

ibilux commented:
> Taking this as customer satisfaction 😅

Yes, customer satisfaction achieved 😂. I was actually planning to start a discussion about multi-DB last month, but I didn’t want to rush things.

> there's also a question of how to deal with it in the SQL editor. Maybe users should just attach DBs manually, e.g. `ATTACH DATABASE get_db_path("other") AS other;` 🤷‍♀️ . Maybe we can provide some UI that lets users select a subset of pre-attached DBs 🤷‍♀️.

Both approaches are viable, and i think neither is a blocker for moving forward.

> Meanwhile trying to fill the gaps with looking at prior art. That's why it's so fascinating to see what Electric an others have done.
> Despite the extra challenges, I agree that an MVP should probably have more sophisticated connection management to avoid later surprises.

Totally agree, given the long-term implications, having smarter connection management in the MVP feels like the best direction. And yes, referencing how previous multi-DB architectures handled this in real-world systems will definitely help guide the design.

On a side note (_just brainstorming here, I realize this might introduce breaking changes_), what if the API structure itself included the database name? Something like:
```js
/api/records/v2/<db_name>/<api_name>
```

This would allow each database to behave like TrailBase does now, but with its own “silo” of attached DBs under the hood. The directory structure might then evolve into something like:
 ```
<traildepot>/data/<db_name>/main.db
<traildepot>/data/<db_name>/migrations/
<traildepot>/data/<db_name>/…
<traildepot>/data/<db_name>/<attached_db_name1>.db
<traildepot>/data/<db_name>/<attached_db_name2>.db
<traildepot>/data/<db_name>/<attached_db_name…>.db
 ```
